contrib/zip/sandboxed.h

```diff
@@ -0,0 +1,52 @@
+// Copyright 2022 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef CONTRIB_ZIP_SANDBOXED_H_
+#define CONTRIB_ZIP_SANDBOXED_H_
+
+#include <libgen.h>
+#include <syscall.h>
+
+#include <memory>
+
+// XXX: This header is required because of the bug in generator.
+// The generator for some reason doesn't catch the types
+// defined by zip (for example zip_uint32_t).
+#include <zipconf.h>
+
+#include "sapi_zip.sapi.h"  // NOLINT(build/include)
+
+class ZipSapiSandbox : public ZipSandbox {
+ public:
+  std::unique_ptr<sandbox2::Policy> ModifyPolicy(
+      sandbox2::PolicyBuilder*) override {
+    return sandbox2::PolicyBuilder()
+        .AllowDynamicStartup()
+        .AllowRead()
+        .AllowWrite()
+        .AllowSystemMalloc()
+        .AllowExit()
+        .AllowSafeFcntl()
+        .AllowSyscalls({
+            __NR_dup,
+            __NR_recvmsg,
+            __NR_ftruncate,
+        })
+        .AllowOpen()
```

Comment:
Can we use `BlockSyscallWithErrno()` here to return `ENOENT`? There should be no reason libzip should read `/etc/localtime`.
I may be wrong, though.

Reply:
This bypass seems to work.
The `/etc/localtime` is read to set a modification time inside the libzip:
https://github.com/nih-at/libzip/blob/6cb36530deafc731cac277080a319d52e4233867/lib/zip_dirent.c#L371

The `_zip_d2u_time` uses `mktime` which read `/etc/localtime`